src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -575,5 +575,53 @@ private static IEnumerable<PropertyInfo> GetAllProperties(Type type)
 
             return allProperties;
         }
+
+        private static object GetPropertyValue(PropertyInfo property, object instance, IConfiguration config, BinderOptions options)
+        {
+            string propertyName = GetPropertyName(property);
+            return BindInstance(
+                property.PropertyType,
+                property.GetValue(instance),
+                config.GetSection(propertyName),
+                options);
+        }
+
+        private static string GetPropertyName(MemberInfo property)
+        {
+            if (property == null)
+            {
+                throw new ArgumentNullException(nameof(property));
+            }
+
+            // Check for a custom property name used for configuration key binding
+            foreach (var attributeData in property.GetCustomAttributesData())
+            {
+                if (attributeData.AttributeType != typeof(ConfigurationKeyNameAttribute) ||
+                    attributeData.ConstructorArguments.Count == 0)
+                {
+                    // ConfigurationKeyName expected to have a single string constructor argument
+                    continue;
+                }
+
+                var constructorArgument = attributeData.ConstructorArguments.First();
+                if (constructorArgument.ArgumentType != typeof(string) ||
+                    constructorArgument.Value == null)
+                {
+                    // First argument didn't contain the expected string value
+                    break;
+                }
+
+                string name = constructorArgument.Value.ToString();
+                if (!string.IsNullOrWhiteSpace(name))
+                {
+                    return name;
+                }
+
+                // No custom string "name" provided - we're done here.
+                break;
```

Comment:
I think we should check the arguments count in a separate if and break, the reasoning for that is because we can only have one `ConfigurationKeyNameAttribute` so in that case if no arguments are provided we will keep looping on all attributes and there is no way we will find another `ConfigurationKeyNameAttribute`. 
```suggestion
                if (attributeData.AttributeType == typeof(ConfigurationKeyNameAttribute))
                {
                    var constructorArgument = attributeData.ConstructorArguments.FirstOrDefault();
                    if (constructorArgument == null ||
                        constructorArgument.ArgumentType != typeof(string) ||
                        constructorArgument.Value == null)
                    {
                        // First argument didn't contain the expected string value
                        break;
                    }

                    string name = constructorArgument.Value.ToString();
                    if (string.IsNullOrWhiteSpace(name))
                    {
                        break;
                    }

                    return name;
                }
```

How about something like this?